Bottom: 2c9034e52b8531a199fe935eb2cc84417480cf26
Top:    707dd6e442696cd42c51f57adf989f482ed735b5
Author: Mark de Wever <koraq@xs4all.nl>
Date:   2021-08-07 11:44:09 +0200

Add a calculator controller object.


---

diff --git a/src/modules/calculator/CMakeLists.txt b/src/modules/calculator/CMakeLists.txt
index 089f030..dc073b6 100644
--- a/src/modules/calculator/CMakeLists.txt
+++ b/src/modules/calculator/CMakeLists.txt
@@ -1,4 +1,5 @@
 add_library(calculator
+	controller.cpp
 	model.cpp
 	value.cpp
 )
@@ -6,9 +7,11 @@ target_compile_options(calculator
 	PRIVATE
 		${module_compile_options}
 )
+add_module(calculator.controller controller.cpp)
 add_module(calculator.model model.cpp)
 add_module(calculator.value value.cpp)
 add_dependencies(calculator
+	calculator.controller.pcm
 	calculator.model.pcm
 	calculator.value.pcm
 )
diff --git a/src/modules/calculator/controller.cpp b/src/modules/calculator/controller.cpp
new file mode 100644
index 0000000..6f36385
--- /dev/null
+++ b/src/modules/calculator/controller.cpp
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) Mark de Wever <koraq@xs4all.nl>
+ * Part of the RPN project.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY.
+ *
+ * See the COPYING file for more details.
+ */
+
+export module calculator.controller;
+
+import calculator.model;
+
+import<charconv>;
+import<format>;
+import<string>;
+import<string_view>;
+
+namespace calculator {
+
+/**
+ * The controller of the calculator.
+ *
+ * For now a very basic version of this class.
+ *
+ * The class is non-copyable and non-movable. These operations make little
+ * sense for the model and are most likely logic errors.
+ */
+export class tcontroller final {
+public:
+  // *** SMF ***
+
+  explicit tcontroller(tmodel &model) noexcept : model_(model) {}
+  tcontroller(const tcontroller &) = delete;
+  tcontroller(tcontroller &&) = delete;
+  ~tcontroller() = default;
+
+  tcontroller &operator=(const tcontroller &) = delete;
+  tcontroller &operator=(tcontroller &&) = delete;
+
+  // *** Operations ***
+  void push() noexcept;
+
+private:
+  void push(std::string_view input);
+
+  void duplicate_last_entry();
+  void parse(std::string_view input);
+
+  tmodel &model_;
+};
+
+void tcontroller::push() noexcept {
+  try {
+    push(model_.input_get());
+    model_.diagnostics_clear();
+  } catch (const std::exception &e) {
+#if defined(__cpp_lib_format)
+    model_.diagnostics_set(std::format("[ERR]{:>80.79}", e.what()));
+#else
+    model_.diagnostics_set(e.what());
+#endif
+  }
+  model_.input_clear();
+}
+
+void tcontroller::push(std::string_view input) {
+  if (input.empty())
+    duplicate_last_entry();
+  else
+    parse(input);
+}
+
+void tcontroller::duplicate_last_entry() {
+  tvalue value = model_.stack_pop();
+  model_.stack_push(value);
+  model_.stack_push(value);
+}
+
+static void validate(std::errc ec) {
+  if (ec == std::errc())
+    return;
+
+  switch (ec) {
+  case std::errc::invalid_argument:
+    throw std::domain_error("Invalid numeric value");
+
+  case std::errc::result_out_of_range:
+    throw std::out_of_range("Value outside of the representable range");
+
+  default:
+    // This happens when the implementation behaves outside the specifications.
+    throw std::domain_error("Unexpected error");
+  }
+}
+
+void tcontroller::parse(std::string_view input) {
+  int64_t value;
+  std::from_chars_result result =
+      std::from_chars(input.begin(), input.end(), value);
+
+  validate(result.ec);
+  if (result.ptr != input.end())
+    throw std::domain_error("Invalid numeric value");
+
+  model_.stack_push(value);
+}
+
+} // namespace calculator
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 24dbb5f..88300e8 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,4 +1,5 @@
 add_executable(tests
+	calculator/controller.cpp
 	calculator/model.cpp
 	calculator/value.cpp
 	math.cpp
diff --git a/tests/calculator/controller.cpp b/tests/calculator/controller.cpp
new file mode 100644
index 0000000..7a9f418
--- /dev/null
+++ b/tests/calculator/controller.cpp
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) Mark de Wever <koraq@xs4all.nl>
+ * Part of the RPN project.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY.
+ *
+ * See the COPYING file for more details.
+ */
+
+import calculator.controller;
+
+import calculator.model;
+
+#include <gtest/gtest.h>
+
+namespace calculator {
+
+TEST(controller, constructor) {
+  tmodel model;
+  const tcontroller controller{model};
+  EXPECT_TRUE(noexcept(tcontroller{model}));
+
+  //
+  //
+  //
+  //
+  //
+  // XXX Backport to other tests..
+}
+
+TEST(controller, copy_constructor) {
+  static_assert(!std::is_copy_constructible_v<tcontroller>,
+                "Implement the proper tests.");
+}
+
+TEST(controller, move_constructor) {
+  static_assert(!std::is_move_constructible_v<tcontroller>,
+                "Implement the proper tests.");
+}
+
+TEST(controller, destructor) {
+  tmodel model;
+  EXPECT_TRUE(noexcept(tcontroller{model}.~tcontroller()));
+}
+
+TEST(controller, copy_assignment) {
+  static_assert(!std::is_copy_assignable_v<tcontroller>,
+                "Implement the proper tests.");
+}
+
+TEST(controller, move_assignment) {
+  static_assert(!std::is_move_assignable_v<tcontroller>,
+                "Implement the proper tests.");
+}
+
+TEST(controller, push) {
+  tmodel model;
+  tcontroller controller{model};
+  EXPECT_TRUE(noexcept(controller.push()));
+}
+
+static std::string format_error(const char *message) {
+#if defined(__cpp_lib_format)
+  return std::format("[ERR]{:>80.79}", message);
+#else
+  return message;
+#endif
+}
+
+TEST(controller, push_duplicate_empty_stack) {
+  tmodel model;
+  tcontroller controller{model};
+
+  controller.push();
+  EXPECT_EQ(model.diagnostics_get(), format_error("Stack is empty"));
+  EXPECT_TRUE(model.stack_empty());
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_duplicate_non_empty_stack) {
+  tmodel model;
+  tcontroller controller{model};
+  model.stack_push(42);
+
+  controller.push();
+  EXPECT_TRUE(model.diagnostics_get().empty());
+  EXPECT_EQ(model.stack_size(), 2);
+  EXPECT_EQ(model.stack_pop().get(), 42);
+  EXPECT_EQ(model.stack_pop().get(), 42);
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_value_0) {
+  tmodel model;
+  tcontroller controller{model};
+  model.input_append("0");
+
+  controller.push();
+  EXPECT_TRUE(model.diagnostics_get().empty());
+  EXPECT_EQ(model.stack_size(), 1);
+  EXPECT_EQ(model.stack_pop().get(), 0);
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_value_min) {
+  tmodel model;
+  tcontroller controller{model};
+  const int64_t minimum = -9223372036854775807ll - 1;
+  static_assert(std::numeric_limits<decltype(model.stack_pop().get())>::min() ==
+                    minimum,
+                "Update the test below");
+
+  model.input_append("-9223372036854775808");
+  controller.push();
+  EXPECT_TRUE(model.diagnostics_get().empty());
+  EXPECT_EQ(model.stack_size(), 1);
+  EXPECT_EQ(model.stack_pop().get(), minimum);
+  EXPECT_TRUE(model.input_get().empty());
+
+  model.input_append("-9223372036854775809");
+  controller.push();
+  EXPECT_EQ(model.diagnostics_get(),
+            format_error("Value outside of the representable range"));
+  EXPECT_TRUE(model.stack_empty());
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_value_max) {
+  tmodel model;
+  tcontroller controller{model};
+  const int64_t maximum = 9223372036854775807ll;
+  static_assert(std::numeric_limits<decltype(model.stack_pop().get())>::max() ==
+                    maximum,
+                "Update the test below");
+
+  model.input_append("9223372036854775807");
+  controller.push();
+  EXPECT_TRUE(model.diagnostics_get().empty());
+  EXPECT_EQ(model.stack_size(), 1);
+  EXPECT_EQ(model.stack_pop().get(), maximum);
+  EXPECT_TRUE(model.input_get().empty());
+
+  model.input_append("9223372036854775808");
+  controller.push();
+  EXPECT_EQ(model.diagnostics_get(),
+            format_error("Value outside of the representable range"));
+  EXPECT_TRUE(model.stack_empty());
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_invalid_input) {
+  tmodel model;
+  tcontroller controller{model};
+
+  model.input_append("a");
+  controller.push();
+  EXPECT_EQ(model.diagnostics_get(), format_error("Invalid numeric value"));
+  EXPECT_TRUE(model.stack_empty());
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+TEST(controller, push_invalid_input_after_valid_value) {
+  tmodel model;
+  tcontroller controller{model};
+
+  model.input_append("0a");
+  controller.push();
+  EXPECT_EQ(model.diagnostics_get(), format_error("Invalid numeric value"));
+  EXPECT_TRUE(model.stack_empty());
+  EXPECT_TRUE(model.input_get().empty());
+}
+
+} // namespace calculator
